### PR TITLE
postgres: graphical query builder

### PR DESCRIPTION
**Pull request by svenklemm**
This enhances the postgres datasource frontend to add a graphical query builder. This still needs a few more improvements before it can be merged but I'm posting this to get some early feedback/testing. 

#10095

TODO:
- [x] unify quoting
- [x] handle unix timestamp as time column (requires #12891)

image

### CODE REVIEW

**Comment by daniellee** [comment]
Looking good! Here's my feedback so far (you probably know all of this already :wink:):

## Where Clause

For the where clause, at the moment you have to manually quote the value for string values:

image

I saw that Clickhouse had the same thing for their adhoc filters but elasticsearch and InfluxDB filters figure that out automatically. It is currently a bit confusing with the dropdown as the strings are not quoted:

image

What are your thoughts on this?

## Group by Time and Metric

Couldn't get this query to work.
image

When I toggled to raw mode (btw totally amazing that you can toggle to edit mode and keep the query!), the query is slightly wrong:

`SELECT $__timeGroup("createdAt",1m),"value","hostname" AS metric FROM "public"."grafana_metric" WHERE $__timeFilter("createdAt") GROUP BY 1 ORDER BY 1`

Think the group by should be `GROUP BY 1,3`.

## Metric Column field

The dropdown didn't show a list of possible columns to use but it worked with manually entering the column name.

## Removing Stuff

Can't remove a metric name or group bys (except for the first group by).

Can remove `time` but not hostname:
image

Will test more but just to say it again - great work Sven!

**Comment by torkelo** [comment]
This looks (functionally) really great! 

I wish this was not needed as this duplicates one of the more complex and messy parts of of Grafana (InfluxDB query editor), but I do see the value in this for making SQL queries more accessible and easier for people to write.  

I wonder if there is a better approach than using InfluxDB query editor as a reference, one that also supported some form of sub query, but that might be too complicated. Maybe break out a component for the where clause, that component and how it works and is built of metric segments is so messy. 

This will need a lot more work to be mergable as there is a ton of complex code here with zero tests. 

Question: where does there model > sql rendering happen? I can only find that it happens live when frontend queries the data source (in datasource.query method), but that means alerting will not work or will have to duplicate all the model > rendering code (something that was unfortunate for InfluxDB and I want to avoid in future, that is reducing duplication between frontend and backend data source code)

**Comment by svenklemm** [comment]
@daniellee 
Regarding the where clause i think the dropdown suggestions should be returned quoted already so they work out of the box. I think autoquoting where parts will limit flexibility for more advanced use cases.

I'll fix the other stuff you mentioned once i get the aggregate function stuff handled properly. I didnt get it to work yet even with your help. The aggregate functions are fetched and if you add a 2nd query you will see aggregate functions actually available in the database. But maybe 2 levels for the SELECT part is overkill anyway since only the aggregate submenu would have multiple entries.

@torkelo 
There is no other editor that could be used as reference though, is there? 

If the code sets rawSQL on every render call then render only needs to be called when you are actually editing the query and alerts should continue to work. No backend changes are required for this to work.

I'll write tests once I get the aggregation function stuff working properly.

**Comment by TSheahan** [comment]
This pull request makes for a very interesting read, and my organization could really use the feature!

Could anyone clarify what this means, in terms of how things are tracking towards the release of this feature?

**Comment by daniellee** [comment]
@TSheahan If this PR is close to mergeable then we add it to the milestone so that someone from the core team reviews it and merges it. But it is ready when it is ready so it might get moved to the next milestone.

**Comment by daniellee** [comment]
@svenklemm let me know when I should review this again or if you have gotten stuck with some frontend stuff.

**Comment by svenklemm** [comment]
@daniellee unfortunately i was busy with other stuff so i didnt get to work on this very much. I started working on a component similar to query_part to handle all the elements of the query but i havent gotten very far with this yet.

**Comment by svenklemm** [comment]
@torkelo  I've written a component to handle the query parts that can't be handled by a single MetricSegment. I hope this addresses the issues you found with the initial version.
@daniellee I think now is a good time for another look. 
I also have a docker-image of this branch at svenklemm/grafana-dev

**Comment by daniellee** [comment]
@svenklemm I just did a quick test using the postgres docker block included in Grafana and I'm getting errors for the aggregate dropdowns and for the metric name column.

Error: `pq: column "aggkind" does not exist` for this query:

`SELECT DISTINCT proname FROM pg_aggregate INNER JOIN pg_proc ON pg_aggregate.aggfnoid = pg_proc.oid INNER JOIN pg_type ON pg_type.oid=pg_proc.prorettype WHERE pronargs=1 AND typname IN ('float8') AND aggkind='n' ORDER BY 1`

Am I doing something wrong? Or does this query not work for certain versions of postgres?

**Comment by svenklemm** [comment]
@daniellee I developed this against postgres 10 and didnt check for compatibility with old versions yet. That aggregate query needs 9.4 I think while the docker block in grafana uses 9.3. I'll do a pass over these metadata queries to make sure they work with older versions aswell. For testing this its probably best to use postgres 9.6 or 10 for now.

**Comment by svenklemm** [comment]
@daniellee The offending query for aggregate functions is removed now, what problems did you have with the metric column?

File: public/app/plugins/datasource/postgres/meta_query.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
+export class PostgresMetaQuery {
+  constructor(private target, private queryModel) {}
+
+  getOperators(datatype: string) {
+    switch (datatype) {
+      case 'float4':
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/meta_query.ts]: Having metadata returned from API with datatypes like string, numeric could help to get rid of checking specific postgres datatypes here. However, you still need to do it somewhere, but maybe certain parts of this then could be extracted to a common base class.

File: public/app/plugins/datasource/postgres/partials/config.html [hunk 2]
```diff
@@ -41,6 +41,17 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
 <h3 class="page-heading">PostgreSQL details</h3>
 
 <div class="gf-form-group">
+	<div class="gf-form">
+		<span class="gf-form-label width-9">
+			Version
+			<info-popover mode="right-normal" position="top center">
+				This option controls what functions are available in the PostgreSQL query builder.
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/partials/config.html]: > This option controls what functions are available in the PostgreSQL qraphical query builder.

Maybe 
... available when using the PostgreSQL query builder.

File: public/app/plugins/datasource/postgres/partials/query.editor.html [hunk 3]
```diff
@@ -1,43 +1,141 @@
-<query-editor-row query-ctrl="ctrl" can-collapse="false">
-	<div class="gf-form-inline">
-		<div class="gf-form gf-form--grow">
-			<code-editor content="ctrl.target.rawSql" datasource="ctrl.datasource" on-change="ctrl.panelCtrl.refresh()" data-mode="sql">
-			</code-editor>
-		</div>
-	</div>
+<query-editor-row query-ctrl="ctrl" has-text-edit-mode="true">
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/partials/query.editor.html]: This file feels very generic for a SQL query editor and probably something we can create a component for later :+1: 

The exception is the help text, but should be trivial to fix.

File: public/app/plugins/datasource/postgres/postgres_query.ts [hunk 4]
```diff
@@ -0,0 +1,285 @@
+import _ from 'lodash';
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/postgres_query.ts]: This class should also be fairly easy to extract to common base class/component eventually :+1:

File: public/app/plugins/datasource/postgres/query_ctrl.ts [hunk 5]
```diff
@@ -1,12 +1,10 @@
 import _ from 'lodash';
+import appEvents from 'app/core/app_events';
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/query_ctrl.ts]: Some parts from this file could also probably be extract to base class/component eventually :+1:

File: public/app/core/components/sql_part/sql_part.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import _ from 'lodash';
```
Comment (marefr) [inline public/app/core/components/sql_part/sql_part.ts]: This component looks good from a "generic" code point of view :+1:

**Review by marefr — COMMENTED** [review]
Added some comments.

**Problems:**
Trying out our pre-built development dashboards ([here](https://github.com/grafana/grafana/blob/master/devenv/dev-dashboards/datasource_tests_postgres_fakedata.json) and [here](https://github.com/grafana/grafana/blob/master/devenv/dev-dashboards/datasource_tests_postgres_unittest.json)) with postgres version 9.3 generates error for certain panels.

I'm guessing that if rawQuery=true the plugin should behave same way as in current master.

<details>
 <summary>Stacktrace</summary>
```
Panel data error: TypeError: Cannot read property 'escapeLiteral' of undefined
    at push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.quoteLiteral (postgres_query.ts:51)
    at arrayMap (lodash.js:639)
    at Function.map (lodash.js:9554)
    at PostgresQuery.push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.interpolateQueryStr (postgres_query.ts:76)
    at TemplateSrv.push../public/app/features/templating/template_srv.ts.TemplateSrv.formatValue (template_srv.ts:91)
    at template_srv.ts:216
    at String.replace (<anonymous>)
    at TemplateSrv.push../public/app/features/templating/template_srv.ts.TemplateSrv.replace (template_srv.ts:188)
    at PostgresQuery.push../public/app/plugins/datasource/postgres/postgres_query.ts.PostgresQuery.render (postgres_query.ts:93)
    at datasource.ts:51
```
</details>

File: public/app/plugins/datasource/postgres/partials/config.html [hunk 7]
```diff
@@ -41,6 +41,17 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
...
+			</info-popover>
+		</span>
+		<span class="gf-form-select-wrapper">
+			<select class="gf-form-input gf-size-auto" ng-model="ctrl.current.jsonData.postgresVersion" ng-options="f.value as f.name for f in ctrl.postgresVersions"></select>
+		</span>
+	</div>
 	<div class="gf-form">
 		<gf-form-switch class="gf-form" label="TimescaleDB" tooltip="Use TimescaleDB features (e.g., time_bucket) in Grafana" label-class="width-9" checked="ctrl.current.jsonData.timescaledb" switch-class="max-width-6"></gf-form-switch>
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/partials/config.html]: For a new user that don't know what TimescaleDB is I think the description could be better. 

Maybe use similar approach to http datasources which have a Help button, see [code](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/partials/ds_http_settings.html#L32). 

A short description with a link to TimescaleDB github repository would be nice I think.

File: public/app/plugins/datasource/postgres/config_ctrl.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+          this.postgresVersions.push({ name: name, value: version });
+        }
+        this.current.jsonData.postgresVersion = version;
+      });
+    });
+  }
+
+  // the value portion is derived from postgres server_version_num/100
+  postgresVersions = [
+    { name: '9.3', value: 903 },
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply public/app/plugins/datasource/postgres/config_ctrl.ts]: 9.4 is the first version that adds a new feature (ordered set aggregate functions), so anything below 9.4 would be treated the same
the number is  how you can query it from postgres itself (```SELECT current_setting('server_version_num');``` divided by 100, the version being integer also makes version checks easier

[the remark this answers is not in the document]
Comment (marefr) [reply public/app/plugins/datasource/postgres/config_ctrl.ts]: Yes makes sense now, thanks.

File: public/app/plugins/datasource/postgres/plugin.json [hunk 9]
```diff
@@ -18,5 +18,10 @@
 
   "alerting": true,
   "annotations": true,
-  "metrics": true
+  "metrics": true,
+
+  "queryOptions": {
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/plugin.json]: Don't think this is needed right now. The $__interval should work anyway. Think it's for calculating intervals on backend.

File: public/app/plugins/datasource/postgres/partials/config.html [hunk 10]
```diff
@@ -38,6 +38,18 @@ <h3 class="page-heading">PostgreSQL Connection</h3>
 	</div>
 </div>
 
+<div class="gf-form-group">
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/partials/config.html]: This is not used right now so think we can skip this for now. The $__interval should work anyway. Think it's for calculating intervals on backend.

**Review by marefr — CHANGES_REQUESTED** [review]
Editor still seems a bit buggy regarding columns wrapped in quotes and needs to be a bit clever to be ably to support users in a user friendly way.

1. I create a new query using gdev-postgres-ds-tests and get the following fields prefilled:
image

2. I select table=metric_values and I would expect the time column to automatically change to a time column available in that table. Instead column from initial table are shown and generates an error "pq: column "c10_timestamp" does not exist". Also the column selected is still the same as for the initial table - would be nice if the columns was either removed and that a query not was sent to backend until every selection is "valid" or either change selected columns to a column from the current selected table.
3. I select time column="timeInt32". Would expect the $__timeFilter to change to $__unixEpochFilter, but it doesn't. This results in a postgres error "pq: invalid input syntax for integer: "2018-03-15T12:30:00Z"" which is not very clear why that is happening.
4. I manually removes $__timeFilter and adds $__unixEpochFilter. Query works
5. I add a group by time. Query fails again "pq: syntax error at or near "timeInt32"" which is not very clear why that is happening, see screenshot below. Seems like the alias incorrectly gets quote added.

image

6. Remove quotes and the query works and gives a plotted graph.

File: public/app/plugins/datasource/postgres/query_ctrl.ts [hunk 11]
```diff
@@ -54,6 +54,7 @@ export class PostgresQueryCtrl extends QueryCtrl {
       if (this.panelCtrl.panel.type === 'table') {
         this.target.format = 'table';
         this.target.rawSql = 'SELECT 1';
+        this.target.rawQuery = true;
```
[the remark this answers is not in the document]
Comment (svenklemm) [reply public/app/plugins/datasource/postgres/query_ctrl.ts]: Yes query builder expects you to have at least a time column and multiple value columns so i guess it should not be default. Queries it creates will still work but will be in the form of
```SELECT time, value1,value2,... FROM table``` so no wildcards and time is not optional.

[the remark this answers is not in the document]
Comment (marefr) [reply public/app/plugins/datasource/postgres/query_ctrl.ts]: You're correct. Please ignore my earlier comment. No changes needed regarding this.

**Comment by marefr** [comment]
Needs fixes:
1. Metric column suggestion varying varchar data type doesn't show up
2. Handle selected column when changing table

**Review by marefr — APPROVED** [review]

**Pull request merged**